### ISSUE TITLE
[GAZPREA] 2024/25 - Rethink Vectors and Matricies

### ISSUE THREAD

**Issue by JustinMeimar** [cmput415/415-docs#46]
### Problem Overview

Gazprea currently allows for vector/matrix declarations with the size parameterized by arbitrary expressions.

For example:
```
integer x = [4..5][1] - 1; // or any expression.
integer[x] my_vec = [1,2,3];
```

Which makes catching size errors at compile time infeasible. This would not be as problematic if Gazprea vectors had defined behavior for using two vector operands of different sizes. However, the vectors Gazprea behave like static arrays; they can not be operated with vectors of different size.

For example:
```
[1, 2, 3] + [1, 2]   // Compile Time SizeError
v1 + v2               // Runtime SizeError
```
This seems like a design conflict. The permissive initialization conditions encourages creating mystery sized vectors, while the use of such vectors is heavily restricted at runtime. It may be better to invert the design; the class expressions vectors can initialized with could be restricted to those decidable at compile-time, while operating between vectors of different size could be defined (like VCalc vectors). Could runtime size errors be defined out of existence entirely?

### Restricting Initialization Size Expressions

A `contexpr` expression is a subset of `const` expressions which are formed exclusively by arithmetic operations on integer literals or other constant expressions.
 
Some examples of a potential `contexpr` construct. 
```
constexpr integer x = 1 + 2;                       // arithmetic op on integer literals
constexpr integer v1_len = x;                      // use of another constexpr
constexpr integer v2_len =  2*v1_len ;             // mix 
```

### Defining Runtime Behavior for Operating Between Different Sizes Vectors 

One option is to inherit vcalc vector operation semantics. Another is to say `constexpr` is sufficient to mitigate most painpoints. The former could define runtime `SizeError`s out of existence. The later does not, but makes them easier to anticipate. 

### Summary
TLDR: We would like to have strong compile time guarantees and reduce the proliferation of runtime `SizeError`s. Currently the spec encourages the opposite: poor compile time guarantees, with no runtime flexibility. 

The proposal is to introduce a qualifier `constexpr` to restrict vector size initialization expressions.

**Comment by Logan9312** [comment]
If I'm thinking about this correctly, requiring vector init size to be known at compile time would mean we don't need to use malloc at all for Gazprea which would render the memory leak tests useless.

It would make implementing vectors way nicer on the other hand, I know I spent a fair amount of time converting from LLVMArrayType to Malloc when I realized Array Type wouldn't cover dynamic vectors well enough

**Comment by JustinMeimar** [comment]
That is a good point @Logan9312 , another implication is that exclusively allowing vectors with compile time known size significantly reduces the set of interesting programs that can be written in Gazprea. I remembered someone in the workshop mentioning, "how many interesting programs can you write in C without malloc," to which the answer is not many.

So perhaps defining operations to push and pop elements from vectors at runtime, defining how to handle operations with mis-matched sizes (likely using reallocation similar to VCalc), or any other number of constructs could be useful to compensate for restricting the initialization expressions.
